### PR TITLE
Upgrade to Lucene 9.12.1

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit upgrades to Lucene 9.12.1.

Among the bug fixes that Lucene 9.12.1 brings, it also allows easier configuration of the Panama vectorization provider with newer Java versions. Set the `org.apache.lucene.vectorization.upperJavaFeatureVersion` system property to increase the set of Java versions that Panama vectorization will provide optimized implementations for. We'll need to carefully bump this sys property in the Elasticsearch 8.x series along side the JDK upgrade. /cc @rjernst 

The addition of the 9.12.1 index version constant in main is handled by #118557. And of course has the same value as what is given in this PR.

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
Note: the IndexVersion constant will need to be forward ported to _main_.

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine run elasticsearch-ci/release-tests

File: server/src/main/java/org/elasticsearch/index/IndexVersions.java [hunk 1]
```diff
@@ -123,6 +123,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion TIME_BASED_K_ORDERED_DOC_ID_BACKPORT = def(8_520_00_0, Version.LUCENE_9_12_0);
     public static final IndexVersion DEPRECATE_SOURCE_MODE_MAPPER = def(8_521_00_0, Version.LUCENE_9_12_0);
     public static final IndexVersion USE_SYNTHETIC_SOURCE_FOR_RECOVERY_BACKPORT = def(8_522_00_0, Version.LUCENE_9_12_0);
+    public static final IndexVersion UPGRADE_TO_LUCENE_9_12_1 = def(8_523_00_0, Version.LUCENE_9_12_1);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/IndexVersions.java:126]: oh good point, what do we need this for if it's unused, can you remind me?

Comment (ChrisHegarty) [reply server/src/main/java/org/elasticsearch/index/IndexVersions.java:126]: It is "statically unused" - there are no static references. However, it is used reflectively to set the value of `LATEST_DEFINED` - the reflective code walks the declared fields.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/IndexVersions.java:126]: gotcha

**Review by javanna — APPROVED** [review]

**Pull request merged**